electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,178 @@
+import os
+import logging
+import argparse
+import numpy as np
+from tqdm import tqdm
+from collections import defaultdict
+import matplotlib.pyplot as plt
+import MDAnalysis as mda
+from solvation_analysis.solute import Solute
+from solvation_analysis._column_names import *
+from pymatgen.core.structure import Molecule
+from solvation_shell_utils import filter_by_rmse, wrap_positions
+
+logging.basicConfig(level=os.environ.get("LOGLEVEL", "INFO"))
+
+
+def extract_solvation_shells(
+    pdb_file_path: str,
+    save_dir: str,
+    system_name: str,
+    solute_atom: str,
+    min_coord: int,
+    max_coord: int,
+    top_n: int,
+):
+    """
+    Given a MD trajectory in a PDB file, perform a solvation analysis
+    on the specified solute to extract the first solvation shell. For each coordination number in the specified range,
+    extract and save the top_n most diverse snapshots based on a RMSD criterion.
+
+    Args:
+        pdb_file_path: Path to the PDB file containing the MD trajectory
+        save_dir: Directory in which to save extracted solvation shells
+        system_name: Name of the system - used for naming the save directory
+        solute_atom: Name (in the PDB file) of the solute atom type (e.g NA0) with which to perform the solvation analysis
+        min_coord: Minimum coordination number to consider
+        max_coord: Maximum coordination number to consider
+        top_n: Number of snapshots to extract per coordination number.
+    """
+
+    # Create save directory
+    os.makedirs(os.path.join(save_dir, system_name, solute_atom), exist_ok=True)
+
+    # Initialize MDA Universe
+    universe = mda.Universe(pdb_file_path)
+
+    # Add PBC box
+    with open(pdb_file_path) as file:
+        dimension_lines = file.readlines()[1]
+        a = float(dimension_lines.split()[1])
+        b = float(dimension_lines.split()[2])
+        c = float(dimension_lines.split()[3])
+        universe.dimensions = [a, b, c, 90, 90, 90]
+
+    lattices = np.array([a, b, c])[None][None]
+
+    # Choose solute atom
+    solu = universe.select_atoms(f"name {solute_atom}")
+
+    logging.info("Translating atoms to solute center of mass")
+    for ts in tqdm(universe.trajectory):
+        ts.dimensions = universe.dimensions
+        solu_center = solu.center_of_mass(wrap=True)
+        dim = ts.triclinic_dimensions
+        box_center = np.sum(dim, axis=0) / 2
+        universe.atoms.translate(box_center - solu_center)
+
+        universe.atoms.unwrap()
+
+    solvent = universe.atoms - solu
+
+    solv_anal = Solute.from_atoms(solu, {"solvent": solvent}, solute_name=solute_atom)
+
+    # Identify the cutoff for the first solvation shell, based on the MD trajectory
+    logging.info("Running solvation analysis")
+    solv_anal.run()
+
+    # Plot the RDF
+    solv_anal.plot_solvation_radius("solute", "solvent")
+    plt.savefig(os.path.join(save_dir, system_name, solute_atom, "solvation_rdf.png"))
```

Comment:
Is this necessary? I was using it to determine reasonable cutoffs, but I don't imagine we'll be wanting to look at RDFs in high throughput

Reply:
It's fairly trivial in terms of runtime and memory so I included it. I figured it doesn't hurt to have them in the directory as a sanity check, but I can remove it if needed.

Reply:
:shrug: Sure, that's fine.